<a href="https://colab.research.google.com/github/Damilare125/SGA_1.4/blob/main/Copy_of_Sql_Queries_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic SQL Queries: Flights Database
© Stutern Graduate Accelerator

Below is an Entity Relationship Diagram (ERD) of the flights database:

<img src="https://github.com/Explore-AI/Pictures/blob/master/flights_ER_diagram.png?raw=true" width=40%/>

As can be seen from the ERD, the flights database consists of four tables:
- **flights**: all domestic flights in the USA in 2008
- **carriers**: lookup table for all the carriers
- **planes**: lookup table for all the planes 
- **airports**: lookup table for all the airports

For this challenge you will apply the use of Basic SQL queries to gain insight into the flights database. 

## Instructions to Students

This challenge is designed to determine how much you have learned so far and will test your knowledge on basic SQL queries.

#### Kindly write your answers in the answer section by double clicking the cell titled "YOUR ANSWER HERE"

Just incase you don't have the flights.db file, click [here](https://drive.google.com/drive/folders/1zoc3iqg3SMn2817C3ZAWeP2Bu-4FrvGS?usp=sharing) to download and ensure you upload it colab 

Do not add or remove cells in this notebook. Do not edit or remove the `%%sql` comment as it is required to run each cell.

### **FINAL NOTICE, DO NOT IGNORE**

* Make sure you select the right answers by running the right queries
* Once you are satisfied with your work, Kindly share a link to your notebook on **citrone**.
* Ensure access to notebook by changing general access from "Restricted" to "Anyone can view". This can be achieved by clicking the share button beside the comment button at the right hand side of the top menu/nav bar .

###**Thanks!**


**_Good Luck!_**


## Loading the database

Before you start producing basic SQL queries, you need to prepare your SQL environment. You can do this by loading in the magic command `%load_ext sql` as we've seen in previous classes. Next you can go ahead and load in your database. To do this you will need to ensure you have downloaded the `flight.db` [here](https://drive.google.com/drive/folders/1zoc3iqg3SMn2817C3ZAWeP2Bu-4FrvGS?usp=sharing)  

In [4]:
%%sql 

sqlite://////content/flights.db

'Connected: @///content/flights.db'

In [3]:
%load_ext sql

#Exploring the database 

Before we begin this challenge we can go ahead and take a look at the composition of each table in the flights database

In [ ]:
%%sql 

PRAGMA table_info(flights);

 * sqlite://////content/flights.db
Done.


[(0, 'index', 'integer', 0, 'NULL', 0),
 (1, 'Date', 'longtext', 0, None, 0),
 (2, 'DayOfWeek', 'integer', 0, 'NULL', 0),
 (3, 'DepTime', 'double', 0, 'NULL', 0),
 (4, 'CRSDepTime', 'longtext', 0, None, 0),
 (5, 'ArrTime', 'double', 0, 'NULL', 0),
 (6, 'CRSArrTime', 'longtext', 0, None, 0),
 (7, 'UniqueCarrier', 'longtext', 0, None, 0),
 (8, 'FlightNum', 'longtext', 0, None, 0),
 (9, 'TailNum', 'longtext', 0, None, 0),
 (10, 'ActualElapsedTime', 'double', 0, 'NULL', 0),
 (11, 'CRSElapsedTime', 'longtext', 0, None, 0),
 (12, 'AirTime', 'longtext', 0, None, 0),
 (13, 'ArrDelay', 'double', 0, 'NULL', 0),
 (14, 'DepDelay', 'double', 0, 'NULL', 0),
 (15, 'Origin', 'longtext', 0, None, 0),
 (16, 'Dest', 'longtext', 0, None, 0),
 (17, 'Distance', 'integer', 0, 'NULL', 0),
 (18, 'TaxiIn', 'longtext', 0, None, 0),
 (19, 'TaxiOut', 'longtext', 0, None, 0),
 (20, 'Cancelled', 'longtext', 0, None, 0),
 (21, 'CancellationCode', 'longtext', 0, None, 0),
 (22, 'Diverted', 'longtext', 0, None, 0),
 (23, 'CarrierDelay', 'longtext', 0, None, 0),
 (24, 'WeatherDelay', 'longtext', 0, None, 0),
 (25, 'NASDelay', 'longtext', 0, None, 0),
 (26, 'SecurityDelay', 'longtext', 0, None, 0),
 (27, 'LateAircraftDelay', 'longtext', 0, None, 0)]

In [5]:
%%sql 

PRAGMA table_info(carriers);

 * sqlite://////content/flights.db
Done.


[(0, 'Code', 'varchar(10)', 0, 'NULL', 0),
 (1, 'Description', 'varchar(100)', 0, 'NULL', 0)]

In [ ]:
%%sql 

PRAGMA table_info(airports);

 * sqlite://////content/flights.db
Done.


[(0, 'iata', 'varchar(10)', 0, 'NULL', 0),
 (1, 'airport', 'varchar(100)', 0, 'NULL', 0),
 (2, 'city', 'varchar(100)', 0, 'NULL', 0),
 (3, 'state', 'varchar(50)', 0, 'NULL', 0),
 (4, 'country', 'varchar(50)', 0, 'NULL', 0),
 (5, 'lat', 'float(24,0)', 0, 'NULL', 0),
 (6, 'long', 'float(24,0)', 0, 'NULL', 0)]

In [ ]:
%%sql 

PRAGMA table_info(planes);

 * sqlite://////content/flights.db
Done.


[(0, 'tailnum', 'varchar(10)', 0, 'NULL', 0),
 (1, 'type', 'varchar(100)', 0, 'NULL', 0),
 (2, 'manufacturer', 'varchar(100)', 0, 'NULL', 0),
 (3, 'issue_date', 'varchar(100)', 0, 'NULL', 0),
 (4, 'model', 'varchar(50)', 0, 'NULL', 0),
 (5, 'status', 'varchar(100)', 0, 'NULL', 0),
 (6, 'aircraft_type', 'varchar(100)', 0, 'NULL', 0),
 (7, 'engine_type', 'varchar(100)', 0, 'NULL', 0),
 (8, 'year', 'longtext', 0, None, 0)]

## Questions on Basic SQL Queries 

Now that you have an idea of what information each table contains, you can use the given cell below each question to execute your basic SQL queries to find the correct output from the options provided for the multiple choice questions on form.

**Question 1**

How many different carriers are there in total in the database?

**Options:**
- 1 252
- 1491
- 20
- 37

In [ ]:
%%sql
--your code here
SELECT COUNT(*)
FROM carriers

 * sqlite://////content/flights.db
Done.


[(1491,)]

### YOUR ANSWER HERE:1491

**Question 2** 

How long was the longest delay before departure? 

**Options:**
- 1355 min
- 999 min
- 1099 min
- 588 min

In [6]:
%%sql
-- your code here
SELECT MAX(DepDelay)
FROM flights;

 * sqlite://////content/flights.db
Done.


[(1355.0,)]

### YOUR ANSWER HERE:1355 min


**Question 3** 

How many flights departed on the 28th of January 2008?

**Options:**
- 540 908
- 20 149
- 19 495
- 18 056

In [ ]:
%%sql
-- your code here
SELECT DepTime, COUNT(*)
FROM flights
WHERE date = 2008/1/28

 * sqlite://////content/flights.db
Done.


[(None, 0)]

### YOUR ANSWER HERE: 0

**Question 4**
 
 What is the distance between Midway Airport (MDW) and Houston Airport (HOU)?
 
 **Options:**
 - 611 km
 - 972 km
 - 1 121 km
 - 937 km

In [ ]:
%%sql
--your code here
SELECT Distance, COUNT(*)
FROM flights
WHERE Origin = "MDW" AND Dest = "HOU"

 * sqlite://////content/flights.db
Done.


[(937, 184)]

### YOUR ANSWER HERE: 937 km

**Question 5**

Which day of the week had the highest number of cancelled flights?(Note 1 = cancelled , 0 = not cancelled;)


**Options:**
- Monday(1)
- Wednesday(3) 
- Saturday(6) 
- Thursday (4) 

In [ ]:
%%sql
-- your code here
SELECT DISTINCT (DayOfWeek), MAX(Cancelled)
FROM flights

 * sqlite://////content/flights.db
Done.


[(4, '1')]

### YOUR ANSWER HERE: Thursday(4)

**Question 6**

How many airports have the word "International" in their name?

**Options:**
- 110
- 124
- 8
- 2

In [7]:
%%sql
--your code here
SELECT COUNT(*)
FROM airports
WHERE airport = "International"
ORDER BY airport

 * sqlite://////content/flights.db
Done.


[(1,)]

### YOUR ANSWER HERE: 1

**Question 7**

What is the most produced model for the manufacturer "BOEING"?

**Options:**
- 717-200
- 737-7H4
- 757-222
- 737-3H4

In [8]:
%%sql
-- your code here
SELECT MAX(model), COUNT(*) AS manufacturer
FROM planes

 * sqlite://////content/flights.db
Done.


[('VANS AIRCRAFT RV6', 5029)]

### YOUR ANSWER HERE: RV6 5029

**Question 8**

How many planes landed at Los Angeles International Airport? 

**Options:**
- 215 608
- 18 964
- 41 258
- 39 422

In [10]:
%%sql

--your code here
SELECT MAX(city), COUNT(*)
FROM airports
WHERE city != "Los Angeles International Airport"

 * sqlite://////content/flights.db
Done.


[('Zuni', 3364)]

### YOUR ANSWER HERE: 3364